src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                     builder.ElementTypeWithAnnotations = collectionType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Single();
 
                     MethodSymbol getEnumeratorMethod = GetGetEnumeratorMethod(collectionType, diagnostics, isAsync, errorLocationSyntax);
+                    MethodSymbol moveNextMethod = null;
 
                     if ((object)getEnumeratorMethod != null)
                     {
                         MethodSymbol specificGetEnumeratorMethod = getEnumeratorMethod.AsMember(collectionType);
-                        builder.GetEnumeratorMethod = specificGetEnumeratorMethod;
                         TypeSymbol enumeratorType = specificGetEnumeratorMethod.ReturnType;
 
                         if (isAsync)
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
```

Comment:
>Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1); [](start = 28, length = 62)

Why is this a valid assumption to make?

Reply:
From our offline discussion, I'll extract the `IEnumerable`/`IAsyncEnumerable` step out into a separate method and inline getting the methods, so it's clear we're dealing with well-known members matched by signature.

---
In reply to: [549813225](https://github.com/dotnet/roslyn/pull/50105#discussion_r549813225) [](ancestors = 549813225)